src/arc.rs

```diff
@@ -183,47 +183,62 @@ impl PathArc {
     /// [`canonicalize`]: struct.PathAbs.html#method.canonicalize
     /// [`current_dir`]: fn.current_dir.html
     pub fn absolute(&self) -> Result<PathAbs> {
-        let mut components = self.components();
-        let mut stack: Vec<OsString> = Vec::new();
-
-        macro_rules! pop_stack { [] => {{
-            if let None = stack.pop() {
-                return Err(Error::new(
-                    io::Error::new(io::ErrorKind::NotFound, ".. consumed root"),
-                    "resolving absolute",
-                    self.clone(),
-                ));
-            }
-        }}}
-
-        handle_prefix(self, &mut stack, &mut components, false)?;
+        let mut res = current_dir(self)?.canonicalize()?;
```

Comment:
I'm concerned that this algorithm is _extremely_ non performant. I think you are creating an array of size `N` for _each_ index. This causes the performance to be `O(N^2)`.

A much more performant approach is to start with an array of components and build it up, which is `O(N)`. This was one of the benefits of the previous implementation (this is much cleaner code however).

Reply:
I think I mentioned somewhere that my original intent was to use `res.push()` to handle each component, instead of recreating `res` from scratch, however `PathArc` doesn't support `.push()` presumably because the `Arc` makes it quite unergonomic.

I'll rewrite it to build up a raw `PathBuf` instead. That means a little extra boiler-plate to translate `io::Error`s to `path_abs::Error`s, but it means we can build the path in-place.

Reply:
gotcha, I didn't catch the context around that at the time.

I wonder if it would make sense to have a trait `PathExt` which is implemented for `Path` and `PathBuf` which overrides the various methods for better errors. That would allow _anyone_ (including us!) to get these better error messages with the core types. Then we could implement `PathExt` for all of our types. It could also help in discouraging the use of `PathArc` (heck, `PathArc` really _would_ just be a deref'd `PathBuf`).

Reply:
I'd definitely like to sketch out a `PathExt` trait and see what effect that might have on existing APIs. I imagine most functions and methods would take `<P: PathExt>`, and we'd implement `PathExt` for `AsRef<Path>` so most things would Just Work automatically.

Once this PR is squared away, I'll give it a try, probably in the name of #28.

Reply:
ya, go ahead and do it with `PathBuf` even with poor errors. Then we can implement `PathExt` and we will have the good errors again.